cmd/acr/purge.go

```diff
@@ -568,43 +573,51 @@ func dryRunPurge(ctx context.Context, acrClient api.AcrCLIClientInterface, login
 			}
 			return -1, -1, err
 		}
-		// This will act as a set if a key is present then it should not be deleted because it is referenced by a multiarch manifest
-		// that will not be deleted
-		doNotDelete := map[string]bool{}
+		// This set will do the check: if a key is present, then the corresponding manifest
+		// should not be deleted because it is referenced by a multiarch manifest
+		// or otherwise references a subject
+		doNotDelete := set.New[string]()
 		candidatesToDelete := []acr.ManifestAttributesBase{}
-		// Iterate over all manifests to discover multiarchitecture manifests
 		for resultManifests != nil && resultManifests.ManifestsAttributes != nil {
 			manifests := *resultManifests.ManifestsAttributes
 			for _, manifest := range manifests {
-				// If the manifest is manifest list and would not get deleted then mark it's dependant manifests as not deletable.
-				if (*manifest.MediaType == manifestListContentType || *manifest.MediaType == manifestIndexContentType) && (*countMap)[*manifest.Digest] != deletedTags[*manifest.Digest] {
-					var manifestListBytes []byte
-					manifestListBytes, err = acrClient.GetManifest(ctx, repoName, *manifest.Digest)
-					if err != nil {
-						return -1, -1, err
-					}
-					var manifestList multiArchManifest
-					err = json.Unmarshal(manifestListBytes, &manifestList)
-					if err != nil {
-						return -1, -1, err
-					}
-					for _, dependentDigest := range manifestList.Manifests {
-						doNotDelete[dependentDigest.Digest] = true
+				if countMap[*manifest.Digest] != deletedTags[*manifest.Digest] {
+					if mediaTypesWithComplexStructure.Contains(*manifest.MediaType) {
+						var manifestBytes []byte
+						manifestBytes, err = acrClient.GetManifest(ctx, repoName, *manifest.Digest)
+						if err != nil {
+							return -1, -1, err
+						}
+						var cm customManifest
+						if err = json.Unmarshal(manifestBytes, &cm); err != nil {
+							return -1, -1, err
+						}
+						candidatesToDelete, doNotDelete, err = getCandidatesToDelete(manifest, cm, doNotDelete, candidatesToDelete)
+						if err != nil {
+							return -1, -1, err
+						}
+					} else {
+						if manifest.Tags == nil {
+							candidatesToDelete = append(candidatesToDelete, manifest)
+						}
 					}
```

Comment:
This block of code seems not matching the original code. Any justifications?

Reply:
yes. I changed the logic, will explain in a separate comment.